drivers/espi/espi_emul.c

```diff
@@ -17,174 +17,472 @@ LOG_MODULE_REGISTER(espi_emul_ctlr);
+static int espi_emul_api_lpc_write_request(const struct device *dev, enum lpc_peripheral_opcode op,
+					   uint32_t *value)
 {
-	ARG_UNUSED(dev);
-	ARG_UNUSED(op);
-	ARG_UNUSED(data);
+	struct espi_emul_data *data = dev->data;
+	struct espi_emul *host_emul = data->host_emul;
+
+	if ((data->channels_enabled & ESPI_CHANNEL_PERIPHERAL) == 0) {
+		return -EIO;
+	}
+
+	switch (op) {
+		/* Write transactions */
+#ifdef CONFIG_ESPI_EMUL_KBC8042
+	case E8042_WRITE_KB_CHAR:
+		data->kbc8042_dbbout = *value;
+		data->kbc8042_status &= ~KBC8042_STATUS_A2;
+		data->kbc8042_status |= KBC8042_STATUS_OBF;
+		break;
+	case E8042_WRITE_MB_CHAR:
+		data->kbc8042_dbbout = *value;
+		data->kbc8042_status |= KBC8042_STATUS_A2;
+		data->kbc8042_status |= KBC8042_STATUS_OBF;
+		break;
+#endif
+		/* Write transactions without input parameters */
+#ifdef CONFIG_ESPI_EMUL_KBC8042
+	case E8042_RESUME_IRQ:
+		data->kbc8042_irq_enabled = true;
+		espi_emul_send_ibf_irq(dev);
+		espi_emul_send_obe_irq(dev);
+		break;
+	case E8042_PAUSE_IRQ:
+		data->kbc8042_irq_enabled = false;
+		break;
+	case E8042_CLEAR_OBF:
+		data->kbc8042_status &= ~KBC8042_STATUS_OBF;
+		break;
+	case E8042_READ_KB_STS:
+		break;
+	case E8042_SET_FLAG:
+		data->kbc8042_status |= *value;
+		break;
+	case E8042_CLEAR_FLAG:
+		data->kbc8042_status &= ~(*value);
+		break;
+#endif
+	case EACPI_WRITE_CHAR:
+		break;
+	/* ACPI status transactions */
+	case EACPI_WRITE_STS:
+		data->acpi_status = *value;
+		break;
+#ifdef CONFIG_ESPI_PERIPHERAL_CUSTOM_OPCODE
+	case ECUSTOM_HOST_SUBS_INTERRUPT_EN:
+		data->interrupts_en = *value;
+		break;
+#endif
+	default:
+		if (host_emul == NULL) {
+			return -EIO;
+		}
 
-	return -EINVAL;
+		return host_emul->api->write_lpc_request(host_emul, op, value);
+	}
+
+	return 0;
 }
 
-static int espi_emul_send_vwire(const struct device *dev, enum espi_vwire_signal vw, uint8_t level)
+/* Logical Channel 1 APIs */
+static int espi_emul_api_send_vwire(const struct device *dev, enum espi_vwire_signal vw,
+				    uint8_t level)
 {
-	const struct emul_espi_device_api *api;
-	struct espi_emul *emul;
 	struct espi_emul_data *data = dev->data;
+	struct espi_emul *host_emul = data->host_emul;
+	struct espi_event evt;
+	int idx;
 
-	if (!(data->cfg.channel_caps & ESPI_CHANNEL_VWIRE)) {
+	if ((data->channels_enabled & ESPI_CHANNEL_VWIRE) == 0) {
 		return -EIO;
 	}
 
-	emul = espi_emul_find(dev, EMUL_ESPI_HOST_CHIPSEL);
-	if (!emul) {
-		LOG_DBG("espi_emul not found");
-		return -EIO;
-	}
+	idx = espi_emul_host_find_index(data, vw);
+	if (idx != -1) {
+		data->vw_state[idx].level = level;
+
+		if (host_emul != NULL) {
+			evt.evt_type = ESPI_BUS_EVENT_VWIRE_RECEIVED;
+			evt.evt_details = vw;
+			evt.evt_data = level;
+
+			host_emul->api->raise_event(host_emul, evt);
+		}
+	} else {
+		if (host_emul == NULL) {
+			return -EIO;
+		}
 
-	__ASSERT_NO_MSG(emul->api);
-	__ASSERT_NO_MSG(emul->api->set_vw);
-	api = emul->api;
+		return host_emul->api->send_vwire(host_emul, vw, level);
+	}
```

Comment:
nit: it is generally more readable to have the error condition first so that the default case doesn't need to be inside an `if`. For example:

```
if (idx == -1) {
  if (host_emul == NULL) {
    return -EIO;
  }
  return host_emul->api->send_vwire(host_enum, vw, level);
}

data->vw_state[idx].level = level;
...
```